File: db4oj.tests/src/com/db4o/db4ounit/common/acid/CrashSimulatingTestCase.java
/* Copyright (C) 2004 - 2005  db4objects Inc.  http://www.db4o.com */

package com.db4o.db4ounit.common.acid;

import java.io.*;

import com.db4o.*;
import com.db4o.config.*;
import com.db4o.foundation.*;
import com.db4o.foundation.io.*;
import com.db4o.internal.*;
import com.db4o.io.*;
import com.db4o.query.*;

import db4ounit.*;
import db4ounit.extensions.fixtures.*;


// TODO: survives commenting out invocations of IoAdaptedObjectContainer#syncFiles() other
// than the one in LocalTransaction#flushFile()
public class CrashSimulatingTestCase implements TestCase, OptOutCS {	
    
	public static class CrashData {
	    public String _name;	    
	    public CrashData _next;

	    public CrashData(CrashData next_, String name) {
	        _next = next_;
	        _name = name;
	    }
	    
		public String toString() {
			return _name+" -> "+_next;
		}
	}
	
    static final boolean VERBOSE = false;
    
    private boolean hasLockFileThread(){
        if (!Platform4.hasLockFileThread()) {
            return false;
        }
        return ! Platform4.hasNio();
    }
    
    public void testWithCache() throws IOException{
        doTest(true);
    }
    
    public void testWithoutCache() throws IOException{
        doTest(false);
    }

    
    private void doTest(boolean cached) throws IOException{
    	if(hasLockFileThread()){
    		System.out.println("CrashSimulatingTestCase is ignored on platforms with lock file thread.");
    		return;
    	}
    	
        String path = Path4.combine(Path4.getTempPath(), "crashSimulate");
        String fileName = Path4.combine(path, "cs");
        
    	File4.delete(fileName);
    	File4.mkdirs(path);
        
    	createFile(baseConfig(), fileName);
        
        CrashSimulatingIoAdapter crashSimulatingAdapter = new CrashSimulatingIoAdapter(new RandomAccessFileAdapter());
        IoAdapter adapterFactory = cached ? (IoAdapter) new CachedIoAdapter(crashSimulatingAdapter) : crashSimulatingAdapter;
        
        Configuration recordConfig = baseConfig();
        recordConfig.io(adapterFactory);
        
        ObjectContainer oc = Db4o.openFile(recordConfig, fileName);
        
        ObjectSet objectSet = oc.queryByExample(new CrashData(null, "three"));
        oc.delete(objectSet.next());
        
        oc.store(new CrashData(null, "four"));
        oc.store(new CrashData(null, "five"));
        oc.store(new CrashData(null, "six"));
        oc.store(new CrashData(null, "seven"));
        oc.store(new CrashData(null, "eight"));
        oc.store(new CrashData(null, "nine"));
        oc.store(new CrashData(null, "10"));
        oc.store(new CrashData(null, "11"));
        oc.store(new CrashData(null, "12"));
        oc.store(new CrashData(null, "13"));
        oc.store(new CrashData(null, "14"));
        
        oc.commit();
        
        Query q = oc.query();
        q.constrain(CrashData.class);
        objectSet = q.execute();
        while(objectSet.hasNext()){
        	CrashData cData = (CrashData) objectSet.next();
            if( !  (cData._name.equals("10") || cData._name.equals("13")) ){
                oc.delete(cData);
            }
        }
        
        oc.commit();

        oc.close();

        int count = crashSimulatingAdapter.batch.writeVersions(fileName);

        checkFiles(fileName, "R", crashSimulatingAdapter.batch.numSyncs());
        checkFiles(fileName, "W", count);
		if (VERBOSE) {
			System.out.println("Total versions: " + count);
		}
    }

	private Configuration baseConfig() {
		Configuration config = Db4o.newConfiguration();
		config.objectClass(CrashData.class).objectField("_name").indexed(true);
    	config.reflectWith(Platform4.reflectorForType(CrashSimulatingTestCase.class));
        config.bTreeNodeSize(4);
		return config;
	}

	private void checkFiles(String fileName, String infix,int count) {
        for (int i = 1; i <= count; i++) {
            if(VERBOSE){
                System.out.println("Checking " + infix + i);
            }
            String versionedFileName = fileName + infix + i;
            ObjectContainer oc = Db4o.openFile(baseConfig(), versionedFileName);
	        try {
	            if(! stateBeforeCommit(oc)){
	                if(! stateAfterFirstCommit(oc)){
	                    Assert.isTrue(stateAfterSecondCommit(oc));
	                }
	            }
            } finally {
            	oc.close();
            }
        }
    }
    
    private boolean stateBeforeCommit(ObjectContainer oc){
        return expect(oc, new String[] {"one", "two", "three"});
    }
    
    private boolean stateAfterFirstCommit (ObjectContainer oc){
        return expect(oc, new String[] {"one", "two", "four", "five", "six", "seven", "eight", "nine", "10", "11", "12", "13", "14" });
    }
    
    private boolean stateAfterSecondCommit (ObjectContainer oc){
        return expect(oc, new String[] {"10", "13"});
    }
    
    private boolean expect(ObjectContainer container, String[] names){
    	Collection4 expected = new Collection4(names);
        ObjectSet actual = container.query(CrashData.class);
        while (actual.hasNext()){
            CrashData current = (CrashData)actual.next();
            if (! expected.remove(current._name)) {
            	return false;
            }
        }
        return expected.isEmpty();
    }
    
    private void createFile(Configuration config, String fileName) throws IOException{
        ObjectContainer oc = Db4o.openFile(config, fileName);
        try {
        	populate(oc);
        } finally {
        	oc.close();
        }
        File4.copy(fileName, fileName + "0");
    }

	private void populate(ObjectContainer container) {
		for (int i = 0; i < 10; i++) {
            container.store(new Item("delme"));
        }
        CrashData one = new CrashData(null, "one");
        CrashData two = new CrashData(one, "two");
        CrashData three = new CrashData(one, "three");
        container.store(one);
        container.store(two);
        container.store(three);
        container.commit();
        ObjectSet objectSet = container.query(Item.class);
        while(objectSet.hasNext()){
            container.delete(objectSet.next());
        }
	}
	
    public static class Item{
    	
        public String name;
        
        public Item() {
        }
        
        public Item(String name_) {
            this.name = name_;
        }

        public String getName() {
            return name;
        }
        
        public void setName(String name_){
        	name = name_;
        }

    }

	
}

File: db4o.net/Db4objects.Db4o.Tests/Db4objects.Db4o.Tests/Common/Acid/CrashSimulatingTestCase.cs
using System.IO;
using Db4oUnit;
using Db4oUnit.Extensions.Fixtures;
using Db4objects.Db4o;
using Db4objects.Db4o.Foundation;
using Db4objects.Db4o.Foundation.IO;
using Db4objects.Db4o.IO;
using Db4objects.Db4o.Internal;
using Db4objects.Db4o.Query;
using Db4objects.Db4o.Tests.Common.Acid;
using Db4objects.Db4o.Tests.Common.Assorted;

namespace Db4objects.Db4o.Tests.Common.Acid
{
	public class CrashSimulatingTestCase : ITestCase, IOptOutCS
	{
		public string _name;

		public Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase _next;

		internal const bool LOG = false;

		public CrashSimulatingTestCase()
		{
		}

		public CrashSimulatingTestCase(Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase
			 next_, string name)
		{
			_next = next_;
			_name = name;
		}

		private bool HasLockFileThread()
		{
			if (!Platform4.HasLockFileThread())
			{
				return false;
			}
			return !Platform4.HasNio();
		}

		public virtual void Test()
		{
			if (HasLockFileThread())
			{
				Sharpen.Runtime.Out.WriteLine("CrashSimulatingTestCase is ignored on platforms with lock file thread."
					);
				return;
			}
			string path = Path.Combine(Path.GetTempPath(), "crashSimulate");
			string fileName = Path.Combine(path, "cs");
			File4.Delete(fileName);
			System.IO.Directory.CreateDirectory(path);
			Db4oFactory.Configure().BTreeNodeSize(4);
			CreateFile(fileName);
			CrashSimulatingIoAdapter adapterFactory = new CrashSimulatingIoAdapter(new RandomAccessFileAdapter
				());
			Db4oFactory.Configure().Io(adapterFactory);
			IObjectContainer oc = Db4oFactory.OpenFile(fileName);
			IObjectSet objectSet = oc.Get(new Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase
				(null, "three"));
			oc.Delete(objectSet.Next());
			oc.Set(new Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase(null, "four"
				));
			oc.Set(new Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase(null, "five"
				));
			oc.Set(new Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase(null, "six")
				);
			oc.Set(new Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase(null, "seven"
				));
			oc.Set(new Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase(null, "eight"
				));
			oc.Set(new Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase(null, "nine"
				));
			oc.Set(new Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase(null, "10"));
			oc.Set(new Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase(null, "11"));
			oc.Set(new Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase(null, "12"));
			oc.Set(new Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase(null, "13"));
			oc.Set(new Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase(null, "14"));
			oc.Commit();
			IQuery q = oc.Query();
			q.Constrain(typeof(Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase));
			objectSet = q.Execute();
			while (objectSet.HasNext())
			{
				Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase cst = (Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase
					)objectSet.Next();
				if (!(cst._name.Equals("10") || cst._name.Equals("13")))
				{
					oc.Delete(cst);
				}
			}
			oc.Commit();
			oc.Close();
			Db4oFactory.Configure().Io(new RandomAccessFileAdapter());
			int count = adapterFactory.batch.WriteVersions(fileName);
			CheckFiles(fileName, "R", adapterFactory.batch.NumSyncs());
			CheckFiles(fileName, "W", count);
		}

		private void CheckFiles(string fileName, string infix, int count)
		{
			for (int i = 1; i <= count; i++)
			{
				string versionedFileName = fileName + infix + i;
				IObjectContainer oc = Db4oFactory.OpenFile(versionedFileName);
				try
				{
					if (!StateBeforeCommit(oc))
					{
						if (!StateAfterFirstCommit(oc))
						{
							Assert.IsTrue(StateAfterSecondCommit(oc));
						}
					}
				}
				finally
				{
					oc.Close();
				}
			}
		}

		private bool StateBeforeCommit(IObjectContainer oc)
		{
			return Expect(oc, new string[] { "one", "two", "three" });
		}

		private bool StateAfterFirstCommit(IObjectContainer oc)
		{
			return Expect(oc, new string[] { "one", "two", "four", "five", "six", "seven", "eight"
				, "nine", "10", "11", "12", "13", "14" });
		}

		private bool StateAfterSecondCommit(IObjectContainer oc)
		{
			return Expect(oc, new string[] { "10", "13" });
		}

		private bool Expect(IObjectContainer container, string[] names)
		{
			Collection4 expected = new Collection4(names);
			IObjectSet actual = container.Query(typeof(Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase)
				);
			while (actual.HasNext())
			{
				Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase current = (Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase
					)actual.Next();
				if (null == expected.Remove(current._name))
				{
					return false;
				}
			}
			return expected.IsEmpty();
		}

		private void CreateFile(string fileName)
		{
			IObjectContainer oc = Db4oFactory.OpenFile(fileName);
			try
			{
				Populate(oc);
			}
			finally
			{
				oc.Close();
			}
			File4.Copy(fileName, fileName + "0");
		}

		private void Populate(IObjectContainer container)
		{
			for (int i = 0; i < 10; i++)
			{
				container.Set(new SimplestPossibleItem("delme"));
			}
			Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase one = new Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase
				(null, "one");
			Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase two = new Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase
				(one, "two");
			Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase three = new Db4objects.Db4o.Tests.Common.Acid.CrashSimulatingTestCase
				(one, "three");
			container.Set(one);
			container.Set(two);
			container.Set(three);
			container.Commit();
			IObjectSet objectSet = container.Query(typeof(SimplestPossibleItem));
			while (objectSet.HasNext())
			{
				container.Delete(objectSet.Next());
			}
		}

		public override string ToString()
		{
			return _name + " -> " + _next;
		}
	}
}
